File: src/bugbot/clients/_provider.py
"""Shared data models + Protocol for PR providers (Bitbucket, GitHub, …).

The orchestrator (`bugbot.services.review.Reviewer`) talks to whichever
client implements `PullRequestProvider` — it doesn't care which forge the
PR came from. Keep this surface small: every method below is *required*
by the reviewer, and adding a new one obligates every existing provider
to implement it.
"""

from __future__ import annotations

from typing import Protocol

from pydantic import BaseModel


class PullRequest(BaseModel):
    id: int
    title: str
    description: str = ""
    source_branch: str
    destination_branch: str
    # `source_commit` is the HEAD of the PR branch — GitHub requires it on
    # every inline review comment (`commit_id`). Bitbucket ignores it on
    # the wire but we still surface it for parity.
    source_commit: str
    destination_commit: str
    author: str


class InlineComment(BaseModel):
    file: str
    line: int
    body: str
    # GitHub-only: head commit the comment is anchored to. The orchestrator
    # populates it from `PullRequest.source_commit`. Bitbucket's client
    # ignores it.
    commit_id: str | None = None
    # Multi-line suggestion support (GitHub). When set and `start_line !=
    # line`, the GitHub client sends `start_line` + `start_side` so the
    # suggestion block replaces the [start_line, line] range. Bitbucket's
    # client ignores it (Bitbucket has no native suggestion concept).
    start_line: int | None = None


class ExistingComment(BaseModel):
    id: int
    file: str | None
    line: int | None
    content: str


class PullRequestProvider(Protocol):
    """Read+write surface the reviewer needs from any forge."""

    @property
    def workspace(self) -> str: ...  # GitHub: owner; Bitbucket: workspace

    @property
    def repo_slug(self) -> str: ...

    @property
    def username(self) -> str: ...

    @property
    def app_password(self) -> str: ...

    @property
    def clone_host(self) -> str: ...  # e.g. "bitbucket.org" / "github.com"

    def get_pull_request(self, pr_id: int) -> PullRequest: ...

    def get_pull_request_diff(self, pr_id: int) -> str: ...

    def list_comments(self, pr_id: int) -> list[ExistingComment]: ...

    def post_summary_comment(self, pr_id: int, body: str) -> dict: ...

    def post_inline_comment(self, pr_id: int, comment: InlineComment) -> dict: ...

    def close(self) -> None: ...


__all__ = [
    "PullRequest",
    "InlineComment",
    "ExistingComment",
    "PullRequestProvider",
]

File: src/clients/provider.rs
//! Shared PR models + the provider abstraction the reviewer talks to.
//!
//! The Python used a `Protocol`; here we use an `enum` for static dispatch
//! (no `async-trait`, no `dyn`). The reviewer holds a `Provider` and the enum
//! delegates each call to the concrete `BitbucketClient` / `GitHubClient`.
//! Provider-specific extras (GitHub thread replies, reviews, fix push) live on
//! `GitHubClient` directly and are reached by the interactive layer.

use serde::{Deserialize, Serialize};

use crate::clients::bitbucket::BitbucketClient;
use crate::clients::github::GitHubClient;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PullRequest {
    pub id: u64,
    pub title: String,
    #[serde(default)]
    pub description: String,
    pub source_branch: String,
    pub destination_branch: String,
    /// HEAD of the PR branch — GitHub requires it (`commit_id`) on every inline
    /// review comment. Bitbucket ignores it on the wire.
    pub source_commit: String,
    pub destination_commit: String,
    pub author: String,
}

#[derive(Debug, Clone)]
pub struct InlineComment {
    pub file: String,
    pub line: u32,
    pub body: String,
    /// GitHub-only: head commit the comment anchors to.
    pub commit_id: Option<String>,
    /// Multi-line suggestion range start (GitHub). When set and `< line`, the
    /// client sends `start_line` + `start_side`.
    pub start_line: Option<u32>,
}

#[derive(Debug, Clone)]
pub struct ExistingComment {
    pub id: i64,
    pub file: Option<String>,
    pub line: Option<u32>,
    pub content: String,
    /// Login/display-name of the author — used for reply-to-bot detection.
    pub author: Option<String>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ProviderKind {
    Bitbucket,
    GitHub,
}

impl ProviderKind {
    pub fn as_str(self) -> &'static str {
        match self {
            ProviderKind::Bitbucket => "bitbucket",
            ProviderKind::GitHub => "github",
        }
    }
}

/// Static-dispatch wrapper over the concrete PR clients.
pub enum Provider {
    Bitbucket(BitbucketClient),
    GitHub(GitHubClient),
}

impl Provider {
    pub fn kind(&self) -> ProviderKind {
        match self {
            Provider::Bitbucket(_) => ProviderKind::Bitbucket,
            Provider::GitHub(_) => ProviderKind::GitHub,
        }
    }

    pub fn clone_host(&self) -> &str {
        match self {
            Provider::Bitbucket(c) => c.clone_host(),
            Provider::GitHub(c) => c.clone_host(),
        }
    }

    pub fn workspace(&self) -> &str {
        match self {
            Provider::Bitbucket(c) => c.workspace(),
            Provider::GitHub(c) => c.workspace(),
        }
    }

    pub fn repo_slug(&self) -> &str {
        match self {
            Provider::Bitbucket(c) => c.repo_slug(),
            Provider::GitHub(c) => c.repo_slug(),
        }
    }

    /// Username for the git-over-HTTPS clone URL.
    pub fn clone_username(&self) -> &str {
        match self {
            Provider::Bitbucket(c) => c.clone_username(),
            Provider::GitHub(c) => c.clone_username(),
        }
    }

    /// Token/password for the git-over-HTTPS clone URL.
    pub fn clone_token(&self) -> &str {
        match self {
            Provider::Bitbucket(c) => c.clone_token(),
            Provider::GitHub(c) => c.clone_token(),
        }
    }

    pub async fn get_pull_request(&self, pr_id: u64) -> anyhow::Result<PullRequest> {
        match self {
            Provider::Bitbucket(c) => Ok(c.get_pull_request(pr_id).await?),
            Provider::GitHub(c) => Ok(c.get_pull_request(pr_id).await?),
        }
    }

    pub async fn get_pull_request_diff(&self, pr_id: u64) -> anyhow::Result<String> {
        match self {
            Provider::Bitbucket(c) => Ok(c.get_pull_request_diff(pr_id).await?),
            Provider::GitHub(c) => Ok(c.get_pull_request_diff(pr_id).await?),
        }
    }

    pub async fn list_comments(&self, pr_id: u64) -> anyhow::Result<Vec<ExistingComment>> {
        match self {
            Provider::Bitbucket(c) => Ok(c.list_comments(pr_id).await?),
            Provider::GitHub(c) => Ok(c.list_comments(pr_id).await?),
        }
    }

    pub async fn post_summary_comment(&self, pr_id: u64, body: &str) -> anyhow::Result<()> {
        match self {
            Provider::Bitbucket(c) => c.post_summary_comment(pr_id, body).await?,
            Provider::GitHub(c) => c.post_summary_comment(pr_id, body).await?,
        };
        Ok(())
    }

    pub async fn post_inline_comment(&self, pr_id: u64, c: &InlineComment) -> anyhow::Result<()> {
        match self {
            Provider::Bitbucket(bb) => bb.post_inline_comment(pr_id, c).await?,
            Provider::GitHub(gh) => gh.post_inline_comment(pr_id, c).await?,
        };
        Ok(())
    }
}
